addons/website_quote/models/sale_order.py

```diff
@@ -0,0 +1,361 @@
+            values['website_description'] = product.quote_description or product.website_description
+        return values
+
+    @api.model
+    def create(self, values):
+        values = self._inject_quote_description(values)
+        return super(SaleOrderLine, self).create(values)
+
+    @api.multi
+    def write(self, values):
+        values = self._inject_quote_description(values)
+        return super(SaleOrderLine, self).write(values)
+
+    # Take the description on the order template if the product is present in it
+    @api.onchange('product_id')
+    def product_id_change(self):
+        domain = super(SaleOrderLine, self).product_id_change()
+        if self.order_id.template_id:
+            self.name = next((quote_line.name for quote_line in self.order_id.template_id.quote_line if quote_line.product_id.id == self.product_id.id), self.name)
+        return domain
+
+
+class SaleOrder(models.Model):
+    _inherit = 'sale.order'
+
+    def _default_template_id(self):
+        template = self.env.ref('website_quote.website_quote_template_default', raise_if_not_found=False)
+        return template
+
+    access_token = fields.Char('Security Token', required=True, copy=False, default=lambda self: str(uuid.uuid4()))
+    template_id = fields.Many2one('sale.quote.template', 'Quotation Template', readonly=True, states={'draft': [('readonly', False)], 'sent': [('readonly', False)]}, default=_default_template_id)
+    website_description = fields.Html('Description', translate=True)
+    options = fields.One2many('sale.order.option', 'order_id', 'Optional Products Lines', readonly=True, states={'draft': [('readonly', False)], 'sent': [('readonly', False)]}, copy=True)
+    amount_undiscounted = fields.Float('Amount Before Discount', compute='_compute_amount_undiscounted', digits=0)
+    quote_viewed = fields.Boolean('Quotation Viewed')
+    require_payment = fields.Selection([
+        (0, 'Not mandatory on website quote validation'),
+        (1, 'Immediate after website order validation'),
+        (2, 'Immediate after website order validation and save a token'),
+    ], 'Payment', help="Require immediate payment by the customer when validating the order from the website quote")
+
+    def _compute_amount_undiscounted(self):
+        for sale_order in self:
+            total = 0.0
+            for line in sale_order.order_line:
+                total += line.price_subtotal + line.price_unit * ((line.discount or 0.0) / 100.0) * line.product_uom_qty # why is there a discount in a field name amount_undiscounted ??
+            sale_order.amount_undiscounted = total
+
+    @api.multi
+    def open_quotation(self):
+        self.ensure_one()
+        self.write({'quote_viewed': True})
+        return {
+            'type': 'ir.actions.act_url',
+            'target': 'self',
+            'url': '/quote/%s/%s' % (self.id, self.access_token)
+        }
+
+    @api.onchange('template_id')
+    def onchange_template_id(self):
```

Comment:
Side note: onchanges methods should be before CRUD / action / business / tools .